src/rgw/rgw_lua_request.cc

```diff
@@ -441,6 +443,10 @@ struct GrantsMetaTable : public EmptyMetaTable {
       const auto it = map->find(std::string(index));
       ceph_assert(it != map->end());
       next_it = std::next(it);
+      // avoid infinite loop in case of multimap-like structures
+      while (next_it != map->end() && next_it->first == it->first) {
+        next_it = std::next(next_it);
+      }
```

Comment:
isn't this going to skip entries? can we not store a real multimap iterator to avoid these lookups by string?

Reply:
yes, it is going to skip entries. stateless iterator does not work well with multimap.
i can store the iterator pointer as light user data. since lua cannot erase elements from the container, there is no risk for invalidation.
will give it a try.